### DISCUSSION TITLE
Interest in Debian Packaging

### DISCUSSION THREAD

**Discussion by frenchwr** [cesanta/mongoose#2622] [Ask questions, get answers] [answered]
Hello, I'm curious if there is any interest (or if there have been any previous conversations) in packaging mongoose (specifically, libmongoose) as a debian package and making it available via Debian and/or Ubuntu public mirrors? This is something I'm personally interested in taking on as I'm a new engineer with Canonical and looking for a new project to package and publish as a learning exercise.

What do you think? Are there any issues (either technical or non-technical) that may make this challenging?

Great job with an awesome tool, by the way. I first used mongoose about five years ago and have been recommending it to others ever since!

**Comment by cpq** [comment]
@frenchwr thanks!

I don't think it makes sense to package it as a library, since mongoose is a single-file library so a library is essentially `mongoose.{c,h}`. It might make sense to make a package out of a binary - see https://mongoose.ws/binary/ , which is a built version of https://github.com/cesanta/mongoose/tree/master/examples/http-server . It works without configuration, supports SSI off-the-bat, and optionally supports file uploads - which is all great for quick file sharing, and handy for web / game developmemnt.

If you decide to make such a package for Debian/Ubuntu - that would be great. I don't think it needs any participation from us though, my understanding is that everything is handled on the Debian side.

Back in 2000s, when mongoose was still called "shttpd", I packaged it for FreeBSD, though did not continue to maintain that package.

**Reply by frenchwr** [reply]
@cpq packaging the http-server example binary sounds like a fun project. I'll start down that path and keep you posted here. The actual creation of the package will probably be fast but having it approved and published may be a slow process. Let me know if have any thoughts/input before I get started.

A few notes/questions that come to mind:

* Is there a convenient way to only run a subset of tests from `test/unit_test.c` that cover code paths from the http-server example?  
* Are there any other binaries (from `examples/` or otherwise) that you think might be suitable for packaging? 
* Unfortunately there is already a package named `mongoose` (that also happens to distribute a binary called `mongoose`) in Debian/Ubuntu [1], so we'll need to use something like `mongoose-http-server` for the package/binary names.

[1] 
```
$ apt info mongoose
Package: mongoose
Version: 1:7.1.0+dfsg-3
Priority: optional
Section: universe/math
Source: suitesparse
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Debian Science Team <debian-science-maintainers@lists.alioth.debian.org>
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Installed-Size: 84.0 kB
Depends: libc6 (>= 2.34), libgcc-s1 (>= 3.3.1), libstdc++6 (>= 13.1), libsuitesparseconfig7 (>= 1:7.0.1)
Suggests: libsuitesparse-doc
Homepage: http://suitesparse.com
Download-Size: 29.5 kB
APT-Sources: http://us.archive.ubuntu.com/ubuntu mantic/universe amd64 Packages
Description: graph partitioning tool that can quickly compute edge cuts (executable)
 Suitesparse is a collection of libraries for computations involving
 sparse matrices.
 .
 Mongoose is a graph partitioning library that can quickly compute edge cuts in
 arbitrary graph. Given a graph with a vertex set and edge set, an edge cut is
 a partitioning of the graph into two subgraphs that are balanced (contain the
 same number of vertices) and the connectivity between the subgraphs is
 minimized (few edges are in the cut).
 .
 This package contains the standalone executable.

```

**Reply by cpq** [reply]
@frenchwr

Thanks!

1. A subset of tests could be initiated by the `make test` command in the repo root
2. I don't think any other binary would be worth packaging
3. `mongoose-http-server` sounds good

**Reply by frenchwr** [reply]
@cpq I've started on this and wanted to see if you would be open to me adding a new target called `test-http-server` to the project's Makefile to isolate coverage for the HTTP server binary.  I _think_ (please confirm) this would include all the `test_http_*()` functions _except_ `test_http_client()`. 

While all unit tests pass from my local, I'm currently hitting test failures for [`test_http_client()`](https://github.com/cesanta/mongoose/blob/ac435bcb917ad570d7bb5a32827d7b77854c090a/test/unit_test.c#L1249) from our build system due to a firewall rule blocking outbound http traffic (there is a dependency on an external resource: http://cesanta.com).

**Reply by cpq** [reply]
@frenchwr thank you!

Please try this patch:
[local.diff.txt](https://github.com/cesanta/mongoose/files/14468992/local.diff.txt)

And run your unit tests using 
```sh
make test TFLAGS="-DLOCALHOST_ONLY"
```

Let me know if that works for you

**Reply by frenchwr** [reply]
@cpq Yes that works well. Thanks!

**Reply by cpq** [reply, marked as answer]
@frenchwr that change got integrated
